docs/proposals/action_based_roles.md

```diff
@@ -0,0 +1,115 @@
+# Action based roles
+
+## Avoid duplication, hide decision in implementation details.
+
+Each component lives through defined life-cycle phases. At the beginning
+a component is installed. As time passes new version of the component is
+available resulting in component upgrade. Under certain conditions one needs
+to carry additional operations to update successfully such as migration,
+backup or post-verification.
+
+Given each of the operations are applied on the same component, some of the operation
+underlying tasks, variables and other Ansible resources are shared. E.g. templates
+of service files, handlers, sequence of installation or configuration tasks.
+
+Additionally, some operations can be unified into one and the proper behavior hidden
+inside a role as implementation details. E.g. component upgrade can be equivalent
+to component installation up to some differences. The role dealing with the installation
+should then decide if the actual installation or rather an update is required.
+
+In practice, it is natural to create a role that deals with an installation as
+the first step. Later, role that deals with an upgrade is implemented. Not rarely
+followed by implementation of other operations. At the end one can end up
+with a set of roles, each responsible for a single operation. Naturally,
+the layout leads to code duplication and various efforts to minimize it.
+
+## Proposal
+
+In case a set of roles conforms to the following conditions:
+
+* each role is responsible for a distinct operation (e.g. installation, upgrade, migration)
+* all operations are applied on the same component
+* the operations share common code base and Ansible resources
+
+consolidate all roles into new one, providing a mechanism to select specific
+operation to apply when the new role is invoked. Once can either go with
```

Comment:
Cool, I had not thought much about running different actions using the same role but it does make a lot of sense.